In [ ]:
ptbr 0.626~0.71~0.740
ptpt 0.706~0.73~0.784

Outras similaridades: Cos, Ovl, Jacc, Dice

Semantic Similarity

These models find semantically similar sentences within one language or across languages:

**distiluse-base-multilingual-cased-v2**: Multilingual knowledge distilled version of multilingual Universal Sentence Encoder. While the original mUSE model only supports 16 languages, this multilingual knowledge distilled version supports 50+ languages.

**xlm-r-distilroberta-base-paraphrase-v1** - Multilingual version of distilroberta-base-paraphrase-v1, trained on parallel data for 50+ languages.

**xlm-r-bert-base-nli-stsb-mean-tokens**: Produces similar embeddings as the bert-base-nli-stsb-mean-token model. Trained on parallel data for 50+ languages.

**distilbert-multilingual-nli-stsb-quora-ranking** - Multilingual version of distilbert-base-nli-stsb-quora-ranking. Fine-tuned with parallel data for 50+ languages.

**T-Systems-onsite/cross-en-de-roberta-sentence-transformer** - Multilingual model for English an German. [More]

# Imports e métodos necessários

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
import pandas as pd 
import xml.etree.ElementTree as et 

from scipy.stats import pearsonr

def parse_xml(xml_file):
    """Parse xml to pandas dataframe."""
    xtree = et.parse(xml_file)
    xroot = xtree.getroot() 

    df_cols = ['id', 't', 'h', 'similarity']
    rows = []

    for node in xroot:
        id_ = node.attrib.get("id")
        similarity = node.attrib.get("similarity")
        t = node.find("t").text
        h = node.find("h").text

        rows.append({
            "id": id_,
            "t": t, 
            "h": h,
            "similarity": similarity
        })
    return pd.DataFrame(rows, columns=df_cols, dtype=float)

def eval_similarity(pairs_gold, pairs_sys):
    '''
    Evaluate the semantic similarity output of the system against a gold score. 
    Results are printed to stdout.
    '''
    
    gold_values = np.array(pairs_gold)
    sys_values = np.array(pairs_sys)
    pearson = pearsonr(gold_values, sys_values)[0]
    absolute_diff = gold_values - sys_values
    mse = (absolute_diff ** 2).mean()
    
    print()
    print('Similarity evaluation')
    print('Pearson\t\tMean Squared Error')
    print('-------\t\t------------------')
    print('{:7.3f}\t\t{:18.2f}'.format(pearson, mse))

# Carregando os dados

In [3]:
!ls ../data/assin

assin-ptbr-dev.xml   assin-ptbr-train.xml  assin-ptpt-test.xml
assin-ptbr-test.xml  assin-ptpt-dev.xml    assin-ptpt-train.xml


In [4]:
df_ptbr_train = parse_xml('../data/assin/assin-ptbr-train.xml')
df_ptbr_dev = parse_xml('../data/assin/assin-ptbr-dev.xml')
df_ptbr_test = parse_xml('../data/assin/assin-ptbr-test.xml')

df_ptpt_train = parse_xml('../data/assin/assin-ptpt-train.xml')
df_ptpt_dev = parse_xml('../data/assin/assin-ptpt-dev.xml')
df_ptpt_test = parse_xml('../data/assin/assin-ptpt-test.xml')

In [5]:
print(f'assin-ptbr-train: {df_ptbr_train.shape}')
print(f'assin-ptbr-dev: {df_ptbr_dev.shape}')
print(f'assin-ptbr-test: {df_ptbr_test.shape}')
print()
print(f'assin-ptpt-train: {df_ptpt_train.shape}')
print(f'assin-ptpt-dev: {df_ptpt_dev.shape}')
print(f'assin-ptpt-test: {df_ptpt_test.shape}')

assin-ptbr-train: (2500, 4)
assin-ptbr-dev: (500, 4)
assin-ptbr-test: (2000, 4)

assin-ptpt-train: (2500, 4)
assin-ptpt-dev: (500, 4)
assin-ptpt-test: (2000, 4)


In [6]:
df_ptbr_train.head()

id                                                  t  \
0  1.0  A gente faz o aporte financeiro, é como se a e...   
1  2.0  Em 2013, a história de como Walt Disney conven...   
2  3.0  David Silva bateu escanteio, Kompany escalou a...   
3  4.0  Para os ambientalistas, as metas anunciadas pe...   
4  5.0  De acordo com a PM, por volta das 10h30 havia ...   

                                                   h  similarity  
0  Fernando Moraes afirma que não tem vínculo com...        2.00  
1  P.L.Travers era completamente contra a adaptaç...        2.25  
2  David Silva cobrou escanteio, o zagueiro se ap...        3.75  
3  Dilma aproveitou seu discurso ontem, na Confer...        2.75  
4  O protesto encerrou por volta de 12h15 (horári...        2.00

# Testes pt-br

## Sentence-BERT

### pre-trained sbert-base-portuguese

In [7]:
from sentence_transformers import models, SentenceTransformer

model = SentenceTransformer('sbert-test', device='cpu')

In [8]:
t_embeddings = model.encode(df_ptbr_test['t'], show_progress_bar=True)
h_embeddings = model.encode(df_ptbr_test['h'], show_progress_bar=True)

In [9]:
from sklearn.metrics.pairwise import  cosine_similarity

similarities = [5.0 * cosine_similarity([t], [h])[0][0] for t, h in zip(t_embeddings, h_embeddings)]

In [10]:
pairs_gold = df_ptbr_test['similarity'].tolist()
pairs_sys = similarities

In [11]:
eval_similarity(pairs_gold, pairs_sys)


Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.722		              0.65


In [12]:
from scipy.stats import spearmanr
print('Spearman correlation: {:7.3f}'.format(spearmanr(pairs_gold, pairs_sys)[0]))

Spearman correlation:   0.710


## Fine-tuning Sentence-BERT

https://www.sbert.net/docs/training/overview.html

https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/sts/training_stsbenchmark_continue_training.py

### pre-trained sbert-base-portuguese

In [8]:
"""
This example loads the pre-trained SentenceTransformer model 'bert-base-nli-mean-tokens' from the server.
It then fine-tunes this model for some epochs on the STS benchmark dataset.
Note: In this example, you must specify a SentenceTransformer model.
If you want to fine-tune a huggingface/transformers model like bert-base-uncased, see training_nli.py and training_stsbenchmark.py
"""
from torch.utils.data import DataLoader
import math
from sentence_transformers import SentenceTransformer,  SentencesDataset, LoggingHandler, losses, util, InputExample
from sentence_transformers.evaluation import EmbeddingSimilarityEvaluator
from datetime import datetime
import os
import csv
from sentence_transformers import models, SentenceTransformer


for i in range(2, 10 + 1):
    # Read the dataset
    train_batch_size = 16
    num_epochs = 4
    model_save_path = f'../data/sbert_ptbr-nli-{i}'# + datetime.now().strftime("%Y-%m-%d_%H-%M-%S")



    # Load a pre-trained sentence transformer model
    #model = SentenceTransformer(model_name)


    model = SentenceTransformer('sbert-test', device='cpu')

    train_samples = []
    dev_samples = []
    test_samples = []

    for i, row in df_ptbr_train.iterrows():
        inp_example = InputExample(texts=[row['t'], row['h']], label=row['similarity'] / 5)
        train_samples.append(inp_example)

    for i, row in df_ptbr_dev.iterrows():
        inp_example = InputExample(texts=[row['t'], row['h']], label=row['similarity'] / 5)
        dev_samples.append(inp_example)

    for i, row in df_ptbr_test.iterrows():
        inp_example = InputExample(texts=[row['t'], row['h']], label=row['similarity'] / 5)
        test_samples.append(inp_example)



    train_dataset = SentencesDataset(train_samples, model)
    train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=train_batch_size)
    train_loss = losses.CosineSimilarityLoss(model=model)


    evaluator = EmbeddingSimilarityEvaluator.from_input_examples(dev_samples, name='sts-dev')


    # Configure the training. We skip evaluation in this example
    warmup_steps = math.ceil(len(train_dataset) * num_epochs / train_batch_size * 0.1) #10% of train data for warm-up


    # Train the model
    model.fit(train_objectives=[(train_dataloader, train_loss)],
              evaluator=evaluator,
              epochs=num_epochs,
              evaluation_steps=1000,
              warmup_steps=warmup_steps,
              output_path=model_save_path)


    ##############################################################################
    #
    # Load the stored model and evaluate its performance on STS benchmark dataset
    #
    ##############################################################################

    model = SentenceTransformer(model_save_path, device='cpu')
    test_evaluator = EmbeddingSimilarityEvaluator.from_input_examples(test_samples, name='sts-test')
    test_evaluator(model, output_path=model_save_path)
    
    t_embeddings = model.encode(df_ptbr_test['t'])
    h_embeddings = model.encode(df_ptbr_test['h'])

    from sklearn.metrics.pairwise import  cosine_similarity

    similarities = [5.0 * cosine_similarity([t], [h])[0][0] for t, h in zip(t_embeddings, h_embeddings)]

    pairs_gold = df_ptbr_test['similarity'].tolist()
    pairs_sys = similarities

    eval_similarity(pairs_gold, pairs_sys)

    from scipy.stats import spearmanr
    print('Spearman correlation: {:7.3f}'.format(spearmanr(pairs_gold, pairs_sys)[0]))




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.806		              0.30
Spearman correlation:   0.795





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.807		              0.30
Spearman correlation:   0.795





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.806		              0.30
Spearman correlation:   0.794





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.808		              0.30
Spearman correlation:   0.796





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.805		              0.31
Spearman correlation:   0.795





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.808		              0.30
Spearman correlation:   0.796





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.804		              0.31
Spearman correlation:   0.792





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.805		              0.30
Spearman correlation:   0.793





Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.809		              0.30
Spearman correlation:   0.797


Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.805		              0.31
Spearman correlation:   0.794

# Testes pt-pt

## Sentence-BERT

### pre-trained sbert-base-portuguese

In [7]:
from sentence_transformers import models, SentenceTransformer

model = SentenceTransformer('sbert-test', device='cpu')

In [15]:
t_embeddings = model.encode(df_ptpt_test['t'].tolist())
h_embeddings = model.encode(df_ptpt_test['h'].tolist())

In [16]:
from sklearn.metrics.pairwise import  cosine_similarity

similarities = [5.0 * cosine_similarity([t], [h])[0][0] for t, h in zip(t_embeddings, h_embeddings)]

In [17]:
pairs_gold = df_ptpt_test['similarity'].tolist()
pairs_sys = similarities

In [18]:
eval_similarity(pairs_gold, pairs_sys)#739


Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.764		              0.52


In [19]:
from scipy.stats import spearmanr
print('Spearman correlation: {:7.3f}'.format(spearmanr(pairs_gold, pairs_sys)[0]))

Spearman correlation:   0.764


## Fine-tuning Sentence-BERT

https://www.sbert.net/docs/training/overview.html

https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/sts/training_stsbenchmark_continue_training.py

### distiluse-base-multilingual-cased-v2

In [7]:
"""
This example loads the pre-trained SentenceTransformer model 'bert-base-nli-mean-tokens' from the server.
It then fine-tunes this model for some epochs on the STS benchmark dataset.
Note: In this example, you must specify a SentenceTransformer model.
If you want to fine-tune a huggingface/transformers model like bert-base-uncased, see training_nli.py and training_stsbenchmark.py
"""
from torch.utils.data import DataLoader
import math
from sentence_transformers import SentenceTransformer,  SentencesDataset, LoggingHandler, losses, util, InputExample
from sentence_transformers.evaluation import EmbeddingSimilarityEvaluator
from datetime import datetime
import os
import csv
from sentence_transformers import models, SentenceTransformer


for i in range(2, 10 + 1):
    # Read the dataset
    train_batch_size = 32
    num_epochs = 4
    model_save_path = f'../data/sbert_assin-ptpt_nli-{i}'# + datetime.now().strftime("%Y-%m-%d_%H-%M-%S")



    # Load a pre-trained sentence transformer model
    #model = SentenceTransformer(model_name)



    model = SentenceTransformer('sbert-test', device='cpu')

    train_samples = []
    dev_samples = []
    test_samples = []

    for i, row in df_ptpt_train.iterrows():
        inp_example = InputExample(texts=[row['t'], row['h']], label=row['similarity'] / 5)
        train_samples.append(inp_example)

    for i, row in df_ptpt_dev.iterrows():
        inp_example = InputExample(texts=[row['t'], row['h']], label=row['similarity'] / 5)
        dev_samples.append(inp_example)

    for i, row in df_ptpt_test.iterrows():
        inp_example = InputExample(texts=[row['t'], row['h']], label=row['similarity'] / 5)
        test_samples.append(inp_example)



    train_dataset = SentencesDataset(train_samples, model)
    train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=train_batch_size)
    train_loss = losses.CosineSimilarityLoss(model=model)


    evaluator = EmbeddingSimilarityEvaluator.from_input_examples(dev_samples, name='sts-dev')


    # Configure the training. We skip evaluation in this example
    warmup_steps = math.ceil(len(train_dataset) * num_epochs / train_batch_size * 0.1) #10% of train data for warm-up


    # Train the model
    model.fit(train_objectives=[(train_dataloader, train_loss)],
              evaluator=evaluator,
              epochs=num_epochs,
              evaluation_steps=1000,
              warmup_steps=warmup_steps,
              output_path=model_save_path)


    ##############################################################################
    #
    # Load the stored model and evaluate its performance on STS benchmark dataset
    #
    ##############################################################################

    model = SentenceTransformer(model_save_path, device='cpu')
    test_evaluator = EmbeddingSimilarityEvaluator.from_input_examples(test_samples, name='sts-test')
    test_evaluator(model, output_path=model_save_path)
    
    t_embeddings = model.encode(df_ptbr_test['t'], show_progress_bar=True)
    h_embeddings = model.encode(df_ptbr_test['h'], show_progress_bar=True)

    from sklearn.metrics.pairwise import  cosine_similarity

    similarities = [5.0 * cosine_similarity([t], [h])[0][0] for t, h in zip(t_embeddings, h_embeddings)]

    pairs_gold = df_ptbr_test['similarity'].tolist()
    pairs_sys = similarities

    eval_similarity(pairs_gold, pairs_sys)

    from scipy.stats import spearmanr
    print('Spearman correlation: {:7.3f}'.format(spearmanr(pairs_gold, pairs_sys)[0]))



Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.784		              0.35
Spearman correlation:   0.775




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.785		              0.35
Spearman correlation:   0.776




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.785		              0.35
Spearman correlation:   0.778




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.786		              0.35
Spearman correlation:   0.778




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.786		              0.35
Spearman correlation:   0.776




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.785		              0.35
Spearman correlation:   0.776




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.784		              0.35
Spearman correlation:   0.777




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.785		              0.35
Spearman correlation:   0.776




Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.784		              0.35
Spearman correlation:   0.775


Similarity evaluation
Pearson		Mean Squared Error
-------		------------------
  0.786		              0.35
Spearman correlation:   0.776